agent/src/storage/mod.rs

```diff
@@ -1,80 +1,348 @@
-    let store = BlobStore::new(path)?;
+/// Our fully-configured storage object, with all of its details hidden.
+pub static STORAGE: OnceCell<Storage> = OnceCell::new();
+
+/// Initialize our local storage and a proxy option if we have no storage ourselves.
+pub async fn initialize(path: Option<PathBuf>) -> ServalResult<()> {
+    let local = if let Some(blobpath) = path {
+        if let Ok(blob) = BlobStore::new(blobpath) {
+            Some(blob)
+        } else {
+            None
+        }
+    } else {
+        None
+    };
+
+    let bucket = if let Ok(bucket_name) = std::env::var("STORAGE_BUCKET") {
+        let region_provider = RegionProviderChain::first_try(
+            std::env::var("AWS_DEFAULT_REGION").ok().map(Region::new),
+        )
+        .or_default_provider()
+        .or_else(Region::new("us-east-2"));
+        let config = aws_config::from_env().region(region_provider).load().await;
+        let bucket = S3Storage::new(&bucket_name, config)?;
+        log::info!("s3 storage bucket enabled at {bucket_name}");
+        Some(bucket)
+    } else {
+        None
+    };
+
+    let store = Storage::new(bucket, local);
     STORAGE.set(store).unwrap();
     Ok(())
 }
 
-/// The job runner calls this to get a storage implementation appropriate for the node.
-/// If we have local storage, we use that. If we do not, we discover a peer that does have
-/// storage and ask it for the data. We don't cache the peer client because it's both cheap
-/// to discover and we would prefer to get one from our most recent idea of who our peers are.
-pub async fn get_runner_storage() -> Result<Box<dyn RunnerStorage + Send + Sync>, ServalError> {
-    match STORAGE.get() {
-        Some(v) => Ok(Box::new(v)),
-        None => {
-            let mesh = MESH.get().expect("Peer network not initialized!"); // yes, we crash in this case
-            let peers = mesh.peers_with_role(&ServalRole::Storage).await;
-            let iter = peers.iter();
-            for peer in iter {
-                if let Some(addr) = peer.http_address() {
-                    let storage = StorageProxy::new(1, addr);
-                    return Ok(Box::new(storage));
+/// This struct holds all the logic for juggling our three different ways of persisting data.
+///
+/// If it has no storage options configured, it immediately proxies all reads and writes to
+/// a freshly-discovered peer that advertises the role. If it has any storage configured,
+/// it will never attempt to proxy, lest we proxy in infinite loops.
+///
+/// If the operation is a read operation, it tries local storage first then falls back to s3 storage
+/// if that is available. If it's a write operation, it will always try all configured options.
+#[derive(Debug, Clone)]
+pub struct Storage {
+    bucket: Option<S3Storage>,
+    local: Option<BlobStore>,
+}
+
+impl Storage {
+    pub fn new(bucket: Option<S3Storage>, local: Option<BlobStore>) -> Self {
```

Comment:
I'm sort of viewing Storage as being composed of a hierarchy of storage locations; `local` takes precedence for reads, followed by `bucket`. One can imagine a hypothetical `mesh` option slotting in between them.